storage/src/test/java/tech/pegasys/artemis/storage/CombinedChainDataClientTest.java

```diff
@@ -0,0 +1,192 @@
+
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+import com.google.common.primitives.UnsignedLong;
+import java.util.Optional;
+import org.apache.tuweni.bytes.Bytes32;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import tech.pegasys.artemis.datastructures.blocks.BeaconBlock;
+import tech.pegasys.artemis.datastructures.state.BeaconState;
+import tech.pegasys.artemis.datastructures.util.BeaconStateUtil;
+import tech.pegasys.artemis.datastructures.util.DataStructureUtil;
+import tech.pegasys.artemis.util.config.Constants;
+
+class CombinedChainDataClientTest {
+
+  private final ChainStorageClient recentChainData = mock(ChainStorageClient.class);
+  private final HistoricalChainData historicalChainData = mock(HistoricalChainData.class);
+  private final Store store = mock(Store.class);
+  private final CombinedChainDataClient client =
+      new CombinedChainDataClient(recentChainData, historicalChainData);
+
+  private int seed = 242842;
+
+  @BeforeEach
+  public void setUp() {
+    when(recentChainData.getStore()).thenReturn(store);
+    when(recentChainData.getFinalizedEpoch()).thenReturn(UnsignedLong.ZERO);
+  }
+
+  @Test
+  public void getBlockBySlot_shouldReturnEmptyWhenRecentDataHasNoStore() {
+    when(recentChainData.getStore()).thenReturn(null);
+    assertThat(client.getBlockAtSlot(UnsignedLong.ONE, Bytes32.ZERO))
+        .isCompletedWithValue(Optional.empty());
+  }
+
+  @Test
+  public void getBlockBySlot_returnEmptyWhenHeadRootUnknown() {
+    when(store.getBlockState(Bytes32.ZERO)).thenReturn(null);
+    assertThat(client.getBlockAtSlot(UnsignedLong.ONE, Bytes32.ZERO))
+        .isCompletedWithValue(Optional.empty());
+  }
+
+  @Test
+  public void getBlockBySlot_returnEmptyWhenHeadRootUnknownAndSlotFinalized() {
+    final UnsignedLong slot = UnsignedLong.ONE;
+    when(store.getBlockState(Bytes32.ZERO)).thenReturn(null);
+    when(recentChainData.getFinalizedEpoch()).thenReturn(UnsignedLong.valueOf(10));
+
+    assertThat(client.getBlockAtSlot(slot, Bytes32.ZERO)).isCompletedWithValue(Optional.empty());
+  }
+
+  @Test
+  public void getBlockBySlot_returnBlockFromHistoricalDataWhenHeadRootKnownAndSlotFinalized() {
+    final UnsignedLong slot = UnsignedLong.ONE;
+    final BeaconBlock block = block(slot);
+    when(store.getBlockState(Bytes32.ZERO)).thenReturn(beaconState(UnsignedLong.valueOf(100)));
+    when(recentChainData.getFinalizedEpoch()).thenReturn(UnsignedLong.valueOf(10));
+    when(historicalChainData.getFinalizedBlockAtSlot(slot))
+        .thenReturn(completedFuture(Optional.of(block)));
+
+    assertThat(client.getBlockAtSlot(slot, Bytes32.ZERO)).isCompletedWithValue(Optional.of(block));
+  }
+
+  @Test
+  public void getBlockBySlot_returnEmptyWhenFinalizedSlotDidNotHaveABlock() {}
```

Comment:
Empty test here

Reply:
implemented this and a couple more around the difference between exact slot vs block in effect.